[WIP] TELCODOCS-1707: Lifecycle Agent GA (#74025)
openshift/openshift-docs: modules/ztp-image-based-upgrade-prep-resources.adoc

// Module included in the following assemblies:
// * edge_computing/image-based-upgrade/cnf-preparing-for-image-based-upgrade.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-creating-resources-with-ztp_{context}"]
= Preparing resources for the image-based upgrade with GitOps ZTP

The {lcao} needs the following resources prepared for the image-based upgrade:

* OADP resources wrapped in a `ConfigMap` object
* Labeled extra manifests

[id="ztp-image-based-upgrade-creating-backup-resources-with-ztp_{context}"]
== Creating OADP resources for the image-based upgrade with GitOps ZTP

You need to prepare your OADP resources to restore your application after an upgrade.

.Prerequisites

* Provision one or more managed clusters with GitOps ZTP.
* Log in as a user with `cluster-admin` privileges.
* Generate a seed image from a compatible seed cluster.

> * Create a separate partition on the target cluster for the container images that is shared between stateroots. For more information about, see _Additional resources_.

Reviewer: ```suggestion
* Create a separate partition on the target cluster for the container images that is shared between stateroots. For more information, see _Additional resources_.
```

generatorOptions:
  disableNameSuffixHash: true 

patches: <3>
- target:
    group: policy.open-cluster-management.io
    version: v1
    kind: Policy
    name: group-ibu-oadp-cm-policy
  patch: |-
    - op: replace
      path: /spec/policy-templates/0/objectDefinition/spec/object-templates/0/objectDefinition/data
      value: '{{hub copyConfigMapData "ztp-group" "oadp-cm" hub}}'
----
<1> Generates the `oadp-cm-policy`.
<2> Creates the `oadp-cm` `ConfigMap` object on the hub cluster with `Backup` and `Restore` CRs.
<3> Overrides the data field of `ConfigMap` added in `oadp-cm-policy`. A hub template is used to propagate the `oadp-cm` `ConfigMap` to all target clusters.

. Push the changes to your Git repository.

[id="ztp-image-based-upgrade-labeling-extramanifests-with-ztp_{context}"]
== Labeling extra manifests for the image-based upgrade with GitOps ZTP

The {lcao} only extracts extra manifests that are labeled with the `lca.openshift.io/target-ocp-version: <target_version>` label.

.Prerequisites

* Provision one or more managed clusters with GitOps ZTP.
* Log in as a user with `cluster-admin` privileges.
* Generate a seed image from a compatible seed cluster.

> * Create a separate partition on the target cluster for the container images that is shared between stateroots. For more information about, see _Additional resources_.

Reviewer: ```suggestion
* Create a separate partition on the target cluster for the container images that is shared between stateroots. For more information, see _Additional resources_.
```

----

[IMPORTANT]
====
The `kustomization.yaml` file must be located in the same directory structure as shown above to reference the `PlatformBackupRestore.yaml` manifest.
====

The `source-crs/ibu/PlatformBackupRestore.yaml` is provided in the ZTP container image.

.PlatformBackupRestore.yaml
include::snippets/ibu-PlatformBackupRestore.adoc[]

If you use {lvms} to create persistent volumes, you can use the `source-crs/ibu/PlatformBackupRestoreLvms.yaml` provided in the ZTP container image to back up your {lvms} resources.

.PlatformBackupRestoreLvms.yaml
[source,yaml]
----
include::snippets/ibu-PlatformBackupRestoreLvms.yaml[]
----

[NOTE]
====
Depending on your {rh-rhacm} configuration, you might need to include the `v1/secrets/open-cluster-management-agent/open-cluster-management-image-pull-credentials` object for backup.
If your `multiclusterHub` CR has `.spec.imagePullSecret` defined and the secret exists on the `open-cluster-management-agent` namespace in your hub cluster, ensure that the object is included in the `lca.openshift.io/apply-label` annotation. If the secret does not exist, you can remove the object from the `apply-label` annotation.
====
--

> . Create the OADP CRs for cluster-scoped application artifacts in the `openshift-adp` namespace.
> You can store these CRs in separate YAML files or you can consolidate them into one single file, for example `custom-oadp-workload-crs.yaml`, separated by the `---` directive.

Reviewer 2: ```suggestion
. Create the OADP CRs for cluster-scoped application artifacts in the `openshift-adp` namespace in the `source-crs/custom-crs` directory.
```

+
--
.Example cluster-scoped application OADP CRs for LSO and {LVMS}
[source,yaml]
----
include::snippets/ibu-ApplicationClusterScopedBackupRestore.yaml[]
----

[IMPORTANT]
====
The same version of the applications must function on both the current and the target release of {product-title}.
====
--

> . Create the OADP CRs for your namespace-scoped application artifacts.

Reviewer 2: ```suggestion
. Create the OADP CRs for your namespace-scoped application artifacts in the `source-crs/custom-crs` directory.
```

.Procedure

> . Ensure that you created the `openshift-adp` and `ztp-group` namespaces.

Reviewer 2: Procedure 1 not necessary 
Procedure 4 is redundant. It’s covered in P2. 
Procedure 3 for application backup and restore CR should be the examples from “Doc additions”

The whole content in kustomization.yaml is needed for generating oadp cm in target cluster, so the procedure is something like

1.  create oadp-cm configmap through oadp-cm-policy  in a new PGT ibu-upgrade-ranGen.yaml
 ```
     ...
      sourceFiles:
      - fileName: ConfigMapGeneric.yaml
        policyName: "oadp-cm-policy"
        metadata:
          name: oadp-cm
          namespace: openshift-adp
```
2.  add the kustomization.yaml with the following content
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

generators: <1>
- ibu-upgrade-ranGen.yaml

configMapGenerator: <2>
- files:
  - source-crs/ibu/PlatformBackupRestore.yaml
  # - <add-workload-oadp-crs-here>
  name: oadp-cm
  namespace: ztp-group
generatorOptions:
  disableNameSuffixHash: true 

patches: <3>
- target:
    group: policy.open-cluster-management.io
    version: v1
    kind: Policy
    name: group-ibu-oadp-cm-policy
  patch: |-
    - op: replace
      path: /spec/policy-templates/0/objectDefinition/spec/object-templates/0/objectDefinition/data
      value: '{{hub copyConfigMapData "ztp-group" "oadp-cm" hub}}'
```
<1> generates oadp cm policy
<2> create a configmap called oadp-cm on the hub cluster with backup and restore CRs
<3> patch to override the data field of configmap added in oadp-cm-policy with hub template to propogate the oadp-cm configmap to all target clusters
3. push the change to git and etc
